that_depends/providers/base.py

```diff
@@ -3,20 +3,22 @@
 
 
 T = typing.TypeVar("T")
+R = typing.TypeVar("R")
+T_co = typing.TypeVar("T_co", covariant=True)
```

Comment:
Covariance is needed here for cases like:

```
from that_depends import BaseContainer, Provide, inject
from that_depends.providers import Factory, List


class A:
    pass


class B(A):
    pass


class C(A):
    pass


class Container(BaseContainer):
    provider_b = Factory(B)
    provider_c = Factory(C)

    my_providers = List[A](provider_b, provider_c)


@inject
async def main(_providers: list[A] = Provide[Container.my_providers]) -> None:
    pass
```

where `mypy` throws the following error:
```
tests/test_covariance.py:21: error: Argument 1 to "List" has incompatible type "Factory[B]"; expected "AbstractProvider[A]"  [arg-type]
tests/test_covariance.py:21: error: Argument 2 to "List" has incompatible type "Factory[C]"; expected "AbstractProvider[A]"  [arg-type]
Found 2 errors in 1 file (checked 23 source files)
task: Failed to run task "lint": exit status 1
```